v2/routing_test.go

```diff
@@ -419,23 +474,493 @@ func TestDHT_FindProvidersAsync_datastore_error(t *testing.T) {
+
+	valueChan, err := d.SearchValue(ctx, "invalid-key")
+	assert.ErrorContains(t, err, "splitting key")
+	assert.Nil(t, valueChan)
+}
+
+func TestDHT_SearchValue_key_for_unsupported_namespace(t *testing.T) {
+	ctx := kadtest.CtxShort(t)
+	d := newTestDHT(t)
+
+	valueChan, err := d.SearchValue(ctx, "/unsupported/key")
+	assert.ErrorIs(t, err, routing.ErrNotSupported)
+	assert.Nil(t, valueChan)
+}
+
+func TestDHT_SearchValue_stops_with_cancelled_context(t *testing.T) {
+	ctx := kadtest.CtxShort(t)
+	cancelledCtx, cancel := context.WithCancel(ctx)
+	cancel()
+
+	// make sure we don't just stop because we don't know any other DHT server
+	top := NewTopology(t)
+	d1 := top.AddServer(nil)
+	d2 := top.AddServer(nil)
+	top.Connect(ctx, d1, d2)
+
+	valueChan, err := d1.SearchValue(cancelledCtx, "/"+namespaceIPNS+"/some-key")
+	assert.NoError(t, err)
+	assertClosed(t, ctx, valueChan)
+}
+
+func TestDHT_SearchValue_has_record_locally(t *testing.T) {
+	// Test setup:
+	// There is just one other server that returns a valid value.
+	ctx := kadtest.CtxShort(t)
+	clk := clock.New()
+
+	_, priv := newIdentity(t)
+	_, validValue := makeIPNSKeyValue(t, clk, priv, 1, time.Hour)
+	key, betterValue := makeIPNSKeyValue(t, clk, priv, 2, time.Hour)
+
+	top := NewTopology(t)
+	d1 := top.AddServer(nil)
+	d2 := top.AddServer(nil)
+
+	top.Connect(ctx, d1, d2)
+
+	err := d1.putValueLocal(ctx, key, validValue)
+	require.NoError(t, err)
+
+	err = d2.putValueLocal(ctx, key, betterValue)
+	require.NoError(t, err)
+
+	valChan, err := d1.SearchValue(ctx, key)
+	require.NoError(t, err)
+
+	val := readItem(t, ctx, valChan) // from local store
+	assert.Equal(t, validValue, val)
+
+	val = readItem(t, ctx, valChan)
+	assert.Equal(t, betterValue, val)
+
+	assertClosed(t, ctx, valChan)
+}
+
+func TestDHT_SearchValue_offline(t *testing.T) {
+	// Test setup:
+	// There is just one other server that returns a valid value.
+	ctx := kadtest.CtxShort(t)
+	d := newTestDHT(t)
+
+	key, v := makePkKeyValue(t)
+	err := d.putValueLocal(ctx, key, v)
+	require.NoError(t, err)
+
+	valChan, err := d.SearchValue(ctx, key, routing.Offline)
+	require.NoError(t, err)
+
+	val := readItem(t, ctx, valChan)
+	assert.Equal(t, v, val)
+
+	assertClosed(t, ctx, valChan)
+}
+
+func TestDHT_SearchValue_offline_not_found_locally(t *testing.T) {
+	// Test setup:
+	// We are connected to a peer that holds the record but require an offline
+	// lookup. Assert that we don't receive the record
+	ctx := kadtest.CtxShort(t)
+
+	key, v := makePkKeyValue(t)
+
+	top := NewTopology(t)
+	d1 := top.AddServer(nil)
+	d2 := top.AddServer(nil)
+
+	top.Connect(ctx, d1, d2)
+
+	err := d2.putValueLocal(ctx, key, v)
+	require.NoError(t, err)
+
+	valChan, err := d1.SearchValue(ctx, key, routing.Offline)
+	assert.ErrorIs(t, err, routing.ErrNotFound)
+	assert.Nil(t, valChan)
+}
+
+func TestDHT_Bootstrap_no_peers_configured(t *testing.T) {
+	// TIMING: this test is based on timeouts - so might become flaky!
+	ctx := kadtest.CtxShort(t)
```

Comment:
We could run timing based tests for longer. CtxShort is for tests that we expect to complete quickly - either the result is processed quickly or it never happens because something is broken. We could have another helper function that always returns a context with a deadline just before the test deadline so the test has the full allotted time to complete

Reply:
Both of the timing-sensitive tests run very fast. It's just that the state doesn't manifests at the remote peers immediately.